token/token_verify.go

```diff
@@ -187,6 +189,52 @@ func (a AuthCheckImpl) checkLocalIssuer(c *gin.Context, strToken string, expecte
 	return nil
 }
 
+// Verify a token against a registered server's public key
+// The token's subject should identify the server (i.e. server id)
+func (a AuthCheckImpl) checkRegisteredServer(ctx *gin.Context, strToken string, expectedScopes []token_scopes.TokenScope, allScopes bool) error {
+	// First parse token without verification to extract claims
+	token, err := jwt.Parse([]byte(strToken), jwt.WithVerify(false))
+	if err != nil {
+		return errors.Wrap(err, "Invalid JWT")
+	}
+
+	// Extract server identifier from token subject
+	subject := token.Subject()
+	if subject == "" {
+		return errors.New("Token missing subject claim")
+	}
+
+	// Look up the server's public key and metadata based on the subject
+	jwks, err := GetJWKSFromServerID(ctx, subject)
+	if err != nil {
+		return errors.Wrap(err, "Failed to lookup registered server")
+	}
+
+	// Now verify the token with the server's public key
+	parsed, err := jwt.Parse([]byte(strToken), jwt.WithKeySet(jwks))
+	if err != nil {
+		return errors.Wrap(err, "Failed to verify JWT with server's registered public key")
```

Comment:
I'd use the phrase "key set" instead of "public key" -- we should always be assuming there are multiple keys whenever we work with a JWKS.